src/Diagnostics.CompilerHost/Startup.cs

```diff
@@ -38,6 +51,41 @@ public Startup(IConfiguration configuration)
         /// <remarks>This method gets called by the runtime. Use this method to add services to the container.</remarks>
         public void ConfigureServices(IServiceCollection services)
         {
+            string openIdConfigEndpoint = $"{Configuration["AzureAd:AADAuthority"]}/.well-known/openid-configuration";
+            var configManager = new ConfigurationManager<OpenIdConnectConfiguration>(openIdConfigEndpoint, new OpenIdConnectConfigurationRetriever());
+            var config = configManager.GetConfigurationAsync().Result;
+            var issuer = config.Issuer;
+            var signingKeys = config.SigningKeys;
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+                .AddJwtBearer(options =>
+                {
+                    options.TokenValidationParameters = new TokenValidationParameters
+                    {
+                        ValidateAudience = true,
+                        ValidAudience = Configuration["AzureAd:ClientId"],
+                        ValidateIssuer = true,
+                        ValidIssuers = new[] { issuer, $"{issuer}/v2.0" },
+                        ValidateLifetime = true,
+                        RequireSignedTokens = true,
+                        IssuerSigningKeys = signingKeys
+                    };
+
+                    options.Events = new JwtBearerEvents
+                    {
+                        OnTokenValidated = context =>
+                        {
+                            var allowedAppIds = Configuration["AzureAd:AllowedAppIds"].Split(",").ToList();
+                            var claimPrincipal = context.Principal;
+                            var appId = claimPrincipal.Claims.FirstOrDefault(c => c.Type.Equals("appid", StringComparison.CurrentCultureIgnoreCase));
+                            if (appId == null || !allowedAppIds.Exists(p => p.Equals(appId.Value, StringComparison.OrdinalIgnoreCase)))
+                            {
+                                context.Fail($"App Id : {appId.Value} is not in the allowed list of App Ids");
```

Comment:
I haven't used `context.Fail()`. Does this shortcircuit the request pipeline and give a 401 Unauthorized?
Also maybe we shouldn't say `App Id : {appId.Value} is not in the allowed list of App Ids`. I think that is too much info for a hacker to know. I think just a 401 Unauthorized is good enough.

Reply:
Yes it will give 401 authorized, I tried it with a different App id and got 401. and yes, I will change the message